### DISCUSSION TITLE
Tracing syscalls made by Python code

### DISCUSSION THREAD

**Discussion by madhusudancs** [bpftrace/bpftrace#2777] [Q&A]
I have a Python script, more specifically a Django management command invoked via `python3[.8] manage.py my-script`, that unexpectedly creates a child process. I want to know which the function that's leading this child process creation.

Here is my idea. Set a tracepoint on `execve` syscall and print the `ustack()`. Looks something like this:

```
bpftrace -p <PID> -e 'tracepoint:syscalls:sys_enter_execve { printf("%s", ustack(3)) }'
```

The problem is I don't see symbols, just addresses. 

1. Why am I not seeing symbols? Is it because my python binary doesn't contain debug symbols? Does using python-dbg on Ubuntu help?
2. Even if I get symbols, these symbols won't be Python function names right? Just the function entry points executed by the python VM (C code). Since Python exposes USDT markers, is there a way to somehow combine USDT information with tracepoints? Is that even conceptually possible with bpftrace, bcc (or perhaps any eBPF tools?)

On a separate note, I am on aarch64, specifically AWS Graviton nodes, and I am seeing a lot of stap note parse errors that look like this when I try to use usdt probes: `8@x0 4@[x21, #:lo12:.LANCHOR2]`. 

Weird thing is this is not even coming from the python binary, but rather glibc.
It seems this was due to the incorrect STAP_SDT_ARG_CONSTRAINT used while compiling gcc for aarch64 previously. It was recently fixed and glibc>=2.37 does not have these local labels. But unfortunately production/production-like environments take a while to reach to reach this version of glibc. 

What can I do in the meantime to tell bpftrace that I don't care about glibc when tracing Python processes?

**Comment by danobi** [comment]
1. Yeah, probably your python binary does not have symbols. Check `nm $(which python`. I'm not sure if installing the debug package will help. It's possible that it will, but would be helpful if you checked and reported back with results.
2. Right. USDT might help if python interpreter exposes a useful probe. Your best bet is to probably use python's built in tracing / debugging facilities. I'm not sure off the top of my head which one would be useful for this, but IIRC there are quite a few things you can do.
